src/cockpit/jsonutil.py

```diff
@@ -104,3 +108,38 @@ def get_objv(obj: JsonObject, key: str, constructor: Callable[[JsonObject], T])
     def as_objv(value: JsonDocument) -> Sequence[T]:
         return tuple(constructor(typechecked(item, dict)) for item in typechecked(value, list))
     return _get(obj, as_objv, key, ())
+
+
+def create_object(message: 'JsonObject | None', kwargs: JsonObject) -> JsonObject:
+    """Constructs a JSON object based on message and kwargs.
+
+    See print_object() for details."""
+    result = dict(message or {})
+
+    for key, value in kwargs.items():
+        # rewrite '_' (necessary in Python syntax kwargs list) to '-' (idiomatic JSON)
+        json_key = key.replace('_', '-')
+        result[json_key] = value
+
+    return result
+
+
+def print_object(message: 'JsonObject | None', kwargs: JsonObject) -> bytes:
+    """Construct and pretty-print a JSON object based on message and kwargs
+
+    If only message is given, it is pretty-printed, unmodified.  If message is
```

Comment:
Well, "unmodified" -- systemd_ctype's encoder encodes Variants to a t/v object and bytes to base64. AFAICS this is only ever used for channel *data* (in `Channel.send_message()`), and control messages are left unmodified. This feels upside down to me -- isn't it the *data* that we want to pass through unmodified, and we may want to prettify/interpret the *control* messages?

Can it ever happen that we get GVariant objects in data messages ? I don't think there's any way to create them in JS, and in Python you'd need gi.Glib. I figure binary data is a thing, though. So this is using systemd_ctypes' encoder for re-using the binary base64 encoder?

Reply:
This is actually a very interesting point, and it's an unintended positive side-effect of all of this work.  Now that we no longer shove D-Bus data through the same channels as control messages, the control message formatting code no longer needs to be able to handle D-Bus types.  We can implement that at the channel level or (better) at the D-Bus level itself, I suppose.  Then we could completely decouple this low-level code from systemd_ctypes.  That would be a very nice cleanup, indeed.

I'll look into this.

Reply:
My question above was too vague and I suppose I was misled. The [one below](https://github.com/cockpit-project/cockpit/pull/19637#discussion_r1401886366) is probably a better starting point.

Reply:
Your idea here has piqued my curiosity.  I'm kinda interesting in pursuing it.

By and large, the overwhelming user of `Channel.send_message()` is the dbus channel.  It uses it ~20 times, in many cases needing to send D-Bus data.

There are other users of JSON-as-payload (filesystem events, metrics updates, packages webserver) which are using plain data.

I have an idea to add a json encoder class attribute to the Channel class and use it for formatting the JSON-as-payload messages.  The DBus channel could then override this.

I notice just now that the packages HTTP backend sends its status message as payload, but the HTTP backend sends a control message.  I'm going to look into what's going on there.  I'm surprised that they're different.

Reply:
Interesting — this is a protocol difference in `http-stream1` (which `-ws` sends) and `https-stream2` (which is used from `cockpit.js`).